### PR TITLE
Adding Support for `willImplement` and `willExtend`

### PR DESCRIPTION
**Pull request by mamazu**
Adding support for those methods in prophecies. However, this only works with the last instance eg.
```
$object = $this->prophecise(A::class);

$object->willExtend(B::class); // type: A|B
$object->willImplement(C::class); // type A|C

As far as I know there is no way you can specify a return type of a class with a new generic because will extens should return ObjectProphecy<T&U> with T being the original type and U being the type of the parameter.

### CODE REVIEW

**Comment by mamazu** [comment]
Refs #2123

File: lib/Extension/Prophecy/stubs/Prophecy.stub [hunk 1]
```diff
@@ -24,6 +24,42 @@ class ObjectProphecy
...
     {
     }
+
+    /**
+     * Forces double to extend specific class.
+     *
+     * @template U
+     * @param class-string<U> $class
+     *
+     * @return self|T|U
```
Comment (dantleech) [inline lib/Extension/Prophecy/stubs/Prophecy.stub:34]: should be `ObjectProphecy<T>` ?

File: lib/Extension/Prophecy/stubs/Prophecy.stub [hunk 2]
```diff
@@ -24,6 +24,42 @@ class ObjectProphecy
...
+    }
+
+    /**
+     * Forces double to implement specific interface.
+     *
+     * @template U
+     *
+     * @param class-string<U> $interface
+     *
+     * @return self|T|U
```
Comment (dantleech) [inline lib/Extension/Prophecy/stubs/Prophecy.stub:47]: should be `ObjectProphecy<T&U>` and might require more than stubs

**Comment by dantleech** [comment]
you'd also need to add tests here: https://github.com/phpactor/phpactor/blob/d59f228113ad30ff0239e35ed8cdcc7e70eb353c/lib/Extension/Prophecy/Tests/Integration/WorseReflection/ProphecyMemberContextResolverTest.php#L12

File: lib/Extension/Prophecy/stubs/Prophecy.stub (hunk 1 above)
Comment (mamazu) [reply lib/Extension/Prophecy/stubs/Prophecy.stub:34]: Well it should be `ObjectProphecy<T&U>` but as far as I know, we don't support that.

Comment (dantleech) [reply lib/Extension/Prophecy/stubs/Prophecy.stub:34]: Arguably that might be what the issue is about :)

Comment (dantleech) [reply lib/Extension/Prophecy/stubs/Prophecy.stub:34]: And it is "T" as it extends, it is not an intersection

Comment (mamazu) [reply lib/Extension/Prophecy/stubs/Prophecy.stub:34]: Yeah I thought so, then I might need to dig into the code a little more.

**Comment by mamazu** [comment]
@dantleech Could you have a look why the test is failing on an unrelated thing?

**Comment by dantleech** [comment]
restarted, flakey test

**Comment by mamazu** [comment]
Then it's ready to merge.

**Comment by dantleech** [comment]
ok - but this isn't correct is it? it's _not_ a union? or?

**Comment by mamazu** [comment]
Well it should be. If I write this code:
```
$mock = $this->prophizize(A::class);
$mock->willExtend(B::class);
```

Then this means the mock Object has this signiture:

```php
class A extends B {}
```

Which then should mean that we should be able to call all functions from `A` and all functions from `B` (technically only the public ones, but this also goes for A). This is the union of A ∪ B. At least that's what I am reading.

**Comment by dantleech** [comment]
It's not A _or_ B. It's just A

**Comment by mamazu** [comment]
Then I don't know what the correct behaviour would be.

**Comment by dantleech** [comment]
if you have `interface Foo` and (for some reason) want to prophesize a class that extends `Bar` _and_ implements `Foo` then you would `->prophesize(Foo::class)->willExtend(Bar::class)`, It's not a union it's `XXX extends Bar implements Foo`, you would the correct type is `XXX`, not a union.

This missing feature is creating that type dynamically, which isn't that trivial.

**Comment by mamazu** [comment]
Any idea what we should do with this PR?

**Pull request closed without merging**